compiler/src/dotty/tools/dotc/core/TypeOps.scala

```diff
@@ -429,17 +430,22 @@ object TypeOps:
             case _ =>
               emptyRange // should happen only in error cases
           }
-        case tp: ThisType if toAvoid(tp.cls) =>
-          range(defn.NothingType, apply(classBound(tp.cls.classInfo)))
+        case tp: ThisType =>
+          tp
```

Comment:
Could you explain why this change is needed and why it's safe to not do avoidance here?

Reply:
We don't need this, this is an optimization -- the rationale is that ThisType doesn't exist outside of `class`, thus they are always safe in avoidance.

Reply:
OK, I think this makes sense but it'd be good to check with an assert that this never happens in our tests (once the tests pass we can disable the assert for performance), and to add a comment explaining that this is safe.